### PR TITLE
Overhaul mixed radix implementation for better readability, composability, and performance

### PR DESCRIPTION
**Pull request by ejmahler**
Hello :) I have a pretty massive pull request for you.

As a summary, this pull request does 4 things:
- Rewrite the butterfly algorithms to operate on contiguous blocks of data, and to store their required twiddle factors internally. This makes them more readable and composable. Because of this change, I was able to easily write size 6, 7, 8, and 16 butterflies.
- Rewrite the mixed radix implementation to operate on FFTAlgorithm trait objects, rather than only being able to call butterflies and recursively call itself. This allowed me to cover the final remaining non-O(nlogn) case (prime numbers raised to a power). So as a result of this change, all inputs are O(nlogn)
- Allow FFTalgorithm instances to use the spectrum buffer as scratch space. This means absolutely none of the FFTAlgorithm instances need extra scratch space, and thus they're all now immutable. This paves the way for the FFT struct itself being immutable, although that isn't included in this PR.
- Rewrite the planner to take advantage of these changes. For example, it caches FFTAlgorithm instances with Rc<>, so that they can be shared as much as possible, since duplication is never desirable.

And there are some smaller changes centered around improving performance. For one example, raders algorithm is now 5x faster than before, and is *always* faster or the same speed compared to a DFT of the same size.

Another smaller change was defining a trait named "FFTnum" and making it depend on `Copy`, `'static`, etc. This way, if we decide to add a new requirement (Or if you need to add Debug while testing!) you only have to make the change in one place instead of 500.

This PR also adds unit tests and size assertions and removes a few unsafe blocks!

Here are some representative benchmarks from before and after these changes: 
**Before**
```
test complex_composite_20736 ... bench:     599,596 ns/iter (+/- 23,358)
test complex_composite_24028 ... bench:   9,980,252 ns/iter (+/- 823,594)
test complex_composite_24576 ... bench:     618,659 ns/iter (+/- 46,099)
test complex_composite_30270 ... bench:   9,409,755 ns/iter (+/- 421,822)
test complex_composite_32192 ... bench:   8,716,421 ns/iter (+/- 445,642)
test complex_p2_00128        ... bench:       1,104 ns/iter (+/- 57)
test complex_p2_00512        ... bench:       5,202 ns/iter (+/- 306)
test complex_p2_02048        ... bench:      24,650 ns/iter (+/- 1,685)
test complex_p2_08192        ... bench:     114,806 ns/iter (+/- 6,772)
test complex_p2_32768        ... bench:     542,033 ns/iter (+/- 27,249)
test complex_p4_00064        ... bench:         395 ns/iter (+/- 78)
test complex_p4_00256        ... bench:       1,926 ns/iter (+/- 103)
test complex_p4_01024        ... bench:       9,211 ns/iter (+/- 429)
test complex_p4_04096        ... bench:      45,977 ns/iter (+/- 3,248)
test complex_p4_16384        ... bench:     217,455 ns/iter (+/- 14,421)
test complex_p7_00343        ... bench:      15,068 ns/iter (+/- 1,696)
test complex_p7_02401        ... bench:     137,142 ns/iter (+/- 14,354)
test complex_p7_16807        ... bench:   1,224,677 ns/iter (+/- 125,012)
test complex_prime_0151      ... bench:      31,650 ns/iter (+/- 3,840)
test complex_prime_1009      ... bench:     297,357 ns/iter (+/- 15,893)
test complex_prime_2017      ... bench:     657,534 ns/iter (+/- 26,547)
test complex_primepower_160801 ... bench: 326,669,381 ns/iter (+/- 23,152,241)
test complex_primepower_44521  ... bench:  40,254,553 ns/iter (+/- 3,041,075)
```

**After**
```
test complex_composite_20736   ... bench:     593,787 ns/iter (+/- 19,329)
test complex_composite_24028   ... bench:   2,679,677 ns/iter (+/- 139,675)
test complex_composite_24576   ... bench:     467,878 ns/iter (+/- 15,977)
test complex_composite_30270   ... bench:   1,501,999 ns/iter (+/- 191,026)
test complex_composite_32192   ... bench:   3,400,551 ns/iter (+/- 195,677)
test complex_p2_00128          ... bench:         867 ns/iter (+/- 47)
test complex_p2_00512          ... bench:       4,996 ns/iter (+/- 257)
test complex_p2_02048          ... bench:      23,821 ns/iter (+/- 1,425)
test complex_p2_08192          ... bench:     111,162 ns/iter (+/- 5,302)
test complex_p2_32768          ... bench:     525,080 ns/iter (+/- 124,971)
test complex_p4_00064          ... bench:         337 ns/iter (+/- 22)
test complex_p4_00256          ... bench:       1,824 ns/iter (+/- 83)
test complex_p4_01024          ... bench:       8,829 ns/iter (+/- 620)
test complex_p4_04096          ... bench:      43,713 ns/iter (+/- 3,450)
test complex_p4_16384          ... bench:     208,375 ns/iter (+/- 6,283)
test complex_p7_00343          ... bench:       6,184 ns/iter (+/- 235)
test complex_p7_02401          ... bench:      58,666 ns/iter (+/- 1,824)
test complex_p7_16807          ... bench:     529,934 ns/iter (+/- 19,615)
test complex_prime_0151        ... bench:       4,725 ns/iter (+/- 166)
test complex_prime_1009        ... bench:      38,675 ns/iter (+/- 1,228)
test complex_prime_2017        ... bench:     104,383 ns/iter (+/- 3,833)
test complex_primepower_160801 ... bench:  12,346,971 ns/iter (+/- 448,396)
test complex_primepower_44521  ... bench:   3,164,215 ns/iter (+/- 146,381)
```

As you can see, every benchmark is either the same or faster.

As always, i'd love any comments on how to improve this before you accept it.

### CODE REVIEW

**Comment by ejmahler** [comment]
What are your plans WRT this pull request? If it's too big to review I understand, if you need more time I understand, i'd just like to know what your thoughts are.

**Comment by awelkie** [comment]
I'm planning on reviewing it today

File: src/algorithm/good_thomas_algorithm.rs [hunk 1]
```diff
@@ -9,26 +11,22 @@ use algorithm::FFTAlgorithm;
 pub struct GoodThomasAlgorithm<T> {
     width: usize,
     // width_inverse: usize,
-    width_size_fft: Box<FFTAlgorithm<T>>,
+    width_size_fft: Rc<FFTAlgorithm<T>>,
 
     height: usize,
     // height_inverse: usize,
-    height_size_fft: Box<FFTAlgorithm<T>>,
+    height_size_fft: Rc<FFTAlgorithm<T>>,
```
Comment (awelkie) [inline src/algorithm/good_thomas_algorithm.rs:18]: At some point I'd like to enable this library to work without the standard library. Is this heap allocation necessary, or could we get away with using generics? Something like `struct GoodThomasAlgorithm<T, A> where A: FFTAlgorithm<T>`.

File: src/algorithm/good_thomas_algorithm.rs [hunk 2]
```diff
@@ -9,26 +11,22 @@ use algorithm::FFTAlgorithm;
...
-impl<T> GoodThomasAlgorithm<T>
-    where T: Signed + FromPrimitive + Copy
-{
-    pub fn new(n1: usize,
-               n1_fft: Box<FFTAlgorithm<T>>,
-               n2: usize,
-               n2_fft: Box<FFTAlgorithm<T>>)
-               -> Self {
+impl<T: FFTnum> GoodThomasAlgorithm<T> {
+    #[allow(dead_code)]
```
Comment (awelkie) [inline src/algorithm/good_thomas_algorithm.rs:25]: Why is this necessary? What's the dead code?

File: tests/accuracy.rs [hunk 3]
```diff
@@ -126,11 +126,23 @@ fn test_cooley_tukey() {
...
+    //test some specific lengths > 100
+    for &len in &[256, 768] {
+        let signal = random_signal(len);
+        assert!(ct_matches_dft(signal, false), "length = {}", len);
+    }
 }
 
 #[test]
 fn test_cooley_tukey_inverse() {
-    for len in 1..100 {
```
Comment (awelkie) [inline tests/accuracy.rs]: Did Cooley Tukey stop working for length 1?

File: src/algorithm/good_thomas_algorithm.rs (hunk 1 above)
Comment (ejmahler) [reply src/algorithm/good_thomas_algorithm.rs:18]: I've been thinking about this exact issue: The library's design currently isn't great for a memory-constrained system, and doesn't work at all if you want zero allocations.

What you're proposing is a neat idea, and it will work for some situations. For example, i think it would be a big improvement to handle Raders' Algorithm this way. But if we're disallowing trait objects (which require heap allocations) then the type of the "parent" must be known at compile time, which would imply that the size of the FFT we want to compute is known at compile time.

There is an alternative to trait objects though: enums. We could create an enum containing every FFT algorithm, and then write a recursive mixed radix implementation that contains a Vec of those enums. That would still require a heap allocation for the Vec, though, but at least we could avoid Box and Rc.

If you want to avoid heap allocations altogether, I've actually been thinking about making another FFT crate called TinyFFT that eliminates heap allocations. It wouldn't compete with RustFFT on performance, but for memory-constrained environments or for no-std environments it would be a useful alternative.

File: src/algorithm/good_thomas_algorithm.rs (hunk 2 above)
Comment (ejmahler) [reply src/algorithm/good_thomas_algorithm.rs:25]: The entire algorithm. It currently isn't used by the planner, because the MixedRadix struct is faster. I kept it around because it's a useful reference implementation for implementing butterflies. For example, I used it to debug my size-6 butterfly, implementation, which uses the good-thomas algorithm.

It might be best to get rid of it if it's not being used. Or maybe put it behind a #[cfg(test)]? what do you think?

File: tests/accuracy.rs (hunk 3 above)
Comment (ejmahler) [reply tests/accuracy.rs]: I changed it to match the non-inverse test function, which has been at 2 for a long time. But you're absolutely right, I should have changed the other one to be 1 instead of changing this one to be 2. I'll make a commit that fixes this.

File: src/algorithm/good_thomas_algorithm.rs (hunk 1 above)
Comment (awelkie) [reply src/algorithm/good_thomas_algorithm.rs:18]: I'm confused; why does knowing the type of a "parent" require that the FFT size is known at compile time? The `FFTAlgorithm` trait allows for sizes that can vary by instance.

File: src/algorithm/good_thomas_algorithm.rs (hunk 2 above)
Comment (awelkie) [reply src/algorithm/good_thomas_algorithm.rs:25]: I think that since this library supports so many FFT algorithms, it would be good to expose all of the different algorithms to the user in addition to providing the planner. So I think it's fine to keep this in for now. Eventually I'd like to be able to let the user decide to use any of the provided algorithms if they want, including this one.

Comment (ejmahler) [reply src/algorithm/good_thomas_algorithm.rs:25]: I agree, it makes sense to expose the algorithms themselves publically.

File: src/algorithm/good_thomas_algorithm.rs (hunk 1 above)
Comment (ejmahler) [reply src/algorithm/good_thomas_algorithm.rs:18]: Because if you have something like Box<Trait>, that's what's called a trait object: https://doc.rust-lang.org/book/trait-objects.html

Trait objects are very useful because they allow for virtual calls, and virtual calls are what allow us to not know (and not care) whether our FFTAlgorithm instance is MixedRadix or Radix4. But trait objects absolutely must sit behind references or owned pointers.

Owned pointers means Rc or box, which means heap allocation. And references w/ stack allocation means complex lifetime management, and I don't think it's possible to design a library that does this ergonomically.

So for practical purposes, we can't use trait objects without heap allocation. Sad :(

This means that a sub-FFT cannot hide which type of FFT it is from its parent, and at some point, someone needs to hold a stack variable with a concrete, known-at-compile-time type. And it's going to be a crazy type like MixedRadix<MixedRadix<RadersAlgorithm<Radix4>,MixedRadix<Butterfly8, Butterfly7>>, RadersAlgorithm<MixedRadix<Radix4, Butterfly5>>>

Summary: We currently "hide" the type of FFT by putting it behind a trait object, but trait objects for all intents and purposes require heap allocation, so we can't use them if we want to avoid heap allocation.

Comment (ejmahler) [reply src/algorithm/good_thomas_algorithm.rs:18]: I think if we want to avoid heap allocation, the strategy I'd use would be to not have a planner -- have free function called `process_planless` that will recursively pull a factor out of `len` and handle it, pull another factor out of `len` and handle it, etc until len is 1. And it'll do this for every FFT evaluation. This way we don't need to store FFTAlgorithm instances in the first place, so this problem goes away.

Comment (awelkie) [reply src/algorithm/good_thomas_algorithm.rs:18]: Oh, I understand the issue now. Maybe we could use something like [managed](https://github.com/m-labs/rust-managed) to allow users to opt-in to using references? I'm not sure how difficult that would be.

Anyways, that doesn't need to be solved in this PR. I'll open an issue so we can track these ideas

**Comment by awelkie** [comment]
The PR overall looks great. I'm ready to merge this in. Is there anything else you wanted to add to this PR before it's merged?

**Comment by ejmahler** [comment]
No, this is everything I have. I don't have any more cards up my sleeve for performance improvements etc. Many of them didn't work, and everything that did is a part of the PR.

I deliberately tried to keep any breaking API changes out of this PR, but this PR enables some useful ones. I'll make a separate issue to talk about them.
